cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrader.java

```diff
@@ -128,4 +149,116 @@ public void create() throws IOException {
     }
   }
 
+  /**
+   * Upgrades all file sets and all datasets that contain an embedded file set: If the properties contain
+   * an absolute base path, convert that into a relative base path. This is because since 2.8, with the
+   * introduction of namespaces, absolute paths have incorrectly been treated the same as realtive paths,
+   * that is, relative to the namespace's data path. This was fixed in 3.1, but that means that existing
+   * file set with absolute paths now do not point to their data anymore. Therefore the upgrade turns all
+   * absolute paths into relative paths by removing the leading "/", which is equivalent under the new
+   * semantics.
+   */
+  private void upgradeFileSets() throws Exception {
+    final DatasetInstanceMDS mds;
+    try {
+      mds = new DatasetMetaTableUtil(dsFramework).getInstanceMetaTable();
+    } catch (Exception e) {
+      LOG.error("Failed to access Datasets instances meta table.");
+      throw e;
+    }
+    TransactionExecutor executor = transactionExecutorFactory.createExecutor(ImmutableList.of((TransactionAware) mds));
+    executor.execute(new TransactionExecutor.Subroutine() {
+      @Override
+      public void apply() throws Exception {
+        MDSKey key = new MDSKey.Builder().add(DatasetInstanceMDS.INSTANCE_PREFIX).build();
+        Map<MDSKey, DatasetSpecification> specs = mds.listKV(key, DatasetSpecification.class);
+        for (Map.Entry<MDSKey, DatasetSpecification> entry : specs.entrySet()) {
+          DatasetSpecification spec = entry.getValue();
+          String absolutePath = containsFileSetWithAbsolutePath(spec);
+          if (absolutePath != null) {
+            Id.DatasetInstance datasetId = toDatasetId(entry.getKey());
+            LOG.info("Migrating spec for dataset {} because it contains a FileSet with absolute path: {}",
+                     datasetId, absolutePath);
+            DatasetSpecification newSpec = migrateDatasetSpec(spec.getName(), spec);
+            mds.write(entry.getKey(), newSpec);
+          }
+        }
+      }
+    });
+  }
+
+  @VisibleForTesting
+  DatasetSpecification migrateDatasetSpec(DatasetSpecification spec) {
```

Comment:
This function can be removed. Directly use `DatasetSpecification migrateDatasetSpec(String datasetName, DatasetSpecification spec) `